core/src/main/java/com/radixdlt/consensus/bft/RoundUpdate.java

```diff
@@ -65,75 +65,30 @@
 package com.radixdlt.consensus.bft;
 
 import com.radixdlt.consensus.HighQC;
+import com.radixdlt.consensus.event.LocalEvent;
 import java.util.Objects;
 
 /** Represents an internal (local) round update. */
-public final class RoundUpdate {
+public interface RoundUpdate extends LocalEvent {
+  HighQC highQC();
 
-  private final Round currentRound;
-  private final HighQC highQC;
+  BFTValidatorId leader();
 
-  private final BFTValidatorId leader;
-  private final BFTValidatorId nextLeader;
+  BFTValidatorId nextLeader();
 
-  private RoundUpdate(
-      Round currentRound, HighQC highQC, BFTValidatorId leader, BFTValidatorId nextLeader) {
-    this.currentRound = currentRound;
-    this.highQC = highQC;
-    this.leader = leader;
-    this.nextLeader = nextLeader;
-  }
+  Round currentRound();
 
-  public static RoundUpdate create(
+  static RoundUpdate create(
       Round currentRound, HighQC highQC, BFTValidatorId leader, BFTValidatorId nextLeader) {
     Objects.requireNonNull(currentRound);
     Objects.requireNonNull(highQC);
     Objects.requireNonNull(leader);
     Objects.requireNonNull(nextLeader);
 
-    return new RoundUpdate(currentRound, highQC, leader, nextLeader);
-  }
-
-  public HighQC getHighQC() {
-    return highQC;
-  }
-
-  public BFTValidatorId getLeader() {
-    return leader;
-  }
-
-  public BFTValidatorId getNextLeader() {
-    return nextLeader;
-  }
-
-  public Round getCurrentRound() {
-    return currentRound;
-  }
-
-  @Override
-  public String toString() {
-    return String.format(
-        "%s[%s %s leader=%s next=%s]",
-        getClass().getSimpleName(), this.currentRound, this.highQC, this.leader, this.nextLeader);
-  }
-
-  @Override
-  public boolean equals(Object o) {
-    if (this == o) {
-      return true;
-    }
-    if (o == null || getClass() != o.getClass()) {
-      return false;
-    }
-    final RoundUpdate that = (RoundUpdate) o;
-    return Objects.equals(currentRound, that.currentRound)
-        && Objects.equals(highQC, that.highQC)
-        && Objects.equals(leader, that.leader)
-        && Objects.equals(nextLeader, that.nextLeader);
-  }
+    record roundUpdate(
```

Comment:
Records have a public constructor, but here we need validation of the input parameters. So, there are two choices: add validation into the constructor, or use static factory method and plain constructor. Throwing exceptions in the constructor, usually recognized as a bad practice, so the only viable option is the static factory method. Switching to interface + private record naturally addresses these issues and enforces "code against interface" policy.

Reply:
> Throwing exceptions in the constructor, usually recognized as a bad practice

I tried to read more about this being bad practice and I've only found vague suggestions that it results in a half-initialized object, and TBH I still don't see how bad it really is.
(I do agree that leaking a reference to `this` before constructor returns is bad, but I don't see relationship to throwing exceptions... Or am I looking in wrong direction, and constructor-exceptions are a bad practice for other reasons?)

It would be a pity to avoid constructor-exceptions, since they are so convenient with `record`s + their compact constructors :(
TBH from conceptual PoV, preventing successful initialization of a plainly-invalid object seems like a good idea, and the constructor is the most surefire place to achieve it. It also produces least boilerplate I think?

Reply:
Well, while in general throwing exceptions in constructor is acceptable, I see this as a violation of the single responsibility principle: constructor validates parameters AND initializes the object. In this regard, the factory method is a better choice as it performs validation and passes valid parameters to constructors.

Reply:
Another consideration is my general dislike to `new` in the code. Factory methods look cleaner to me, but this is a personal preference.

Reply:
I also favor factory methods, for their flexibility.

I love that e.g. `Set.of(..)` can internally do:
```
if 0 args, return new EmptySet(); // impl optimized for being empty
if 1 arg, return new SingletonSet(arg); // impl optimized for a single elem
if <10 args, return new TreeSet(args); // whatever other optimizations
if >10 args, return new HashSet(args); // ...
```
_(please excuse the silly example with Set, where this approach is probably overkill)_

And the above makes lot of sense when the constructed object is polymorphic, or maybe at least can be customized in non-trivial ways.

However, when everyone expects that the constructed event is just a POJO, and a very rigid one, and there is dozens different event types, then I think it's better to optimize for "simpler syntax, less boilerplate, direct usage of language's constructs" than for "factory method flexibility".

Reply:
I think that factory methods are a better approach from the point of view of consistency (some classes are not records, and it might be hard/inconvenient to make them records) and separation of concerns: the constructor should assemble the object, but validation is a separate task and it better be done in the factory method.

Reply:
It's more about consistency than about flexibility. There are some classes which have more than one static factory method. Also, in some cases, factory methods are used to hide some implementation details and keep defaults in one place. For example, ScheduledLocalTimeout. Now, only convention prevents direct instantiation via constructor, but I'd prefer this expressed directly in the code rather than expect that nobody will do this.

Reply:
I'd personally just go for a top level `record RoundUpdate(...) implements LocalEvent` and ditch that non-null validation. That inner record seems a bit weird/unnecessary to me.
If we really want the validation then I'd go for an additional compact constructor, e.g.:
```
    public RoundUpdate {
        Objects.requireNonNull(...);
        ...
    }
```

Reply:
OK, looks like common consensus is to make all such classes records (where possible, there are some exceptions). Please let me know if there are any objections.

Reply:
sounds cool, I expect it will actually remove a lot of boilerplate 👍 

Regarding the "`requireNonNull(...)` inside compact constructors" - I typically value the fail-fast behavior, so I'd recommend to keep it (for the price of "whoops actually some boilerplate again!").
BUT if veterans @LukasGasior1 and @siy confirm that this codebase never had problems with cryptic NPEs from getters, then we can drop this clearly-overkill non-null validation too.

Reply:
At the best of my knowledge, there were never any NPE's or validation exceptions.

Reply:
hah yeah I think we've got a team of pretty well established never-nullers :smile: